[Search Pipelines] Add stats for search pipelines

msfroh opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
This adds statistics on executions and time spent on search pipeline operations, similar to the stats that are available for ingest pipelines.

### Related Issues
Resolves #6723 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

lukas-vlcek commented on server/src/test/java/org/opensearch/search/pipeline/SearchPipelineStatsTests.java:
```diff
@@ -0,0 +1,183 @@
+        String jsonContent = bos.toString(StandardCharsets.UTF_8);
+
+        assertEquals(
+            "{\n"
```
Isn't there a better way of testing JSON content? To me this seems to be too sensitive to formatting.

msfroh replied:
That's a good point -- depending on the JSON formatting is risky and brittle.

My concerns with recreating the structure with XContentBuilder methods are:
1. To some extent, we're duplicating the logic that we're testing (without the conditionals/loops).
2. The JSON is "arguably" more readable -- I used this test as a confirmation that the response looks how I expected it. (I did want to use a `"""` string block to hide the  escaped quotes + newlines, but I think we're still on language level 11.)

What I managed to do to avoid depending on the exact JSON string is:

```
String expected = "...giant JSON string...";
XContentParser expectedParser = JsonXContent.jsonXContent.createParser(this.xContentRegistry(), DeprecationHandler.THROW_UNSUPPORTED_OPERATION, expected);
XContentBuilder expectedBuilder = XContentBuilder.builder(JsonXContent.jsonXContent);
expectedBuilder.generator().copyCurrentStructure(expectedParser);

assertEquals(XContentHelper.convertToMap(BytesReference.bytes(expectedBuilder), false, (MediaType) XContentType.JSON),
             XContentHelper.convertToMap(BytesReference.bytes(actualBuilder), false, (MediaType) XContentType.JSON));
```

It removes the format sensitivity but still serves as a "preview" of what the JSON looks like. What do you think?

saratvemulapalli commented on server/src/main/java/org/opensearch/search/pipeline/SearchPipelineMetrics.java:
```diff
@@ -0,0 +1,56 @@
+ * compatible open source license.
+ */
+
+package org.opensearch.search.pipeline;
```
`SearchPipelineMetrics` is very similar implementation of `IngestMetric`[1]. Lets consolidate the logic with a base class? 

[1] https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/ingest/IngestMetric.java

msfroh replied:
I think they can actually be one class used in both places. I'll make that change.

saratvemulapalli commented on server/src/main/java/org/opensearch/search/pipeline/Pipeline.java:
```diff
@@ -41,33 +45,55 @@ class Pipeline {
+    private final List<ProcessorWithMetrics<SearchResponseProcessor>> searchResponseProcessors;
 
     private final NamedWriteableRegistry namedWriteableRegistry;
+    private final SearchPipelineMetrics totalRequestMetrics;
+    private final SearchPipelineMetrics totalResponseMetrics;
+    private final SearchPipelineMetrics pipelineRequestMetrics = new SearchPipelineMetrics();
+    private final SearchPipelineMetrics pipelineResponseMetrics = new SearchPipelineMetrics();
```
Metrics logic is really tangled in with this class. 
Personally I would just abstract it out to keep `Pipeline.java` simple.

msfroh replied:
Whatever class invokes a pipeline and the individual processors needs to hold the references to the various metrics (in order to invoke `before`, `after`, and `failed` on the metrics).

I guess I could subclass `Pipeline`, creating a `PipelineWithMetrics`. The base class could define empty methods `beforeTransformRequest`, `afterTransformRequest`, `beforeRequestProcessor(String processorKey)`, `afterRequestProcessor(String processorKey)`, etc. The subclass would hold the metrics instances and delegate to the various metrics instances.

Is that what you had in mind?

saratvemulapalli replied:
Yup I like that, mostly what I was looking for. 
Also I would take it further, defining `request`/`response` processors and letting `Pipeline.java` know about it is not really needed. I would rather pass the Processor and its got its Type as one of the parameters.

msfroh replied:
I took the subclassing approach. I was worried that it would be hairy, but after taking a shot at it, the separation between "transforming stuff" and "tracking / managing metrics" looks (reasonably) clean.

Thanks for the suggestion!

saratvemulapalli reviewed (commented):
Doing a first pass.. couple of comments but overall the changes look good.

msfroh commented:
@saratvemulapalli  -- Addressed your suggestions.

Windows precommit failed, but that seems to be a "Windows build server" thing and not a "this PR" thing.

saratvemulapalli commented:
> @saratvemulapalli -- Addressed your suggestions.
> 
> Windows precommit failed, but that seems to be a "Windows build server" thing and not a "this PR" thing.

Thanks @msfroh. I've triggered a re-run of window precommit. I haven't seen failures just in Windows platform. 
If the failure persists we can dig in.

msfroh commented:
Thanks, @saratvemulapalli! 

I resolved the open conversations, so it looks like there are no immediate merge blockers.

msfroh commented:
Oh, the "Some checks were not successful" note seems to be there because the Mac OS precommit was canceled when the Windows precommit failed.

dblock commented:
@msfroh needs manual backport if you want this in 2.x

mingshl commented:
I will help with the backport to 2.x

msfroh commented:
> I totally get it that given current state of API isolation it is possible that any internal renaming is possibly a "breaking" change but hey ... it is. I do not know if it makes sense to establish a new rule about how the "breaking" label is applied but the fact is that change like this can cause compilation error somewhere else.

Oh, shoot. I didn't expect that.

I'm not sure what the best option is there. I'll aim to add the "breaking" label if I rename an existing method in future, at least.

lukas-vlcek commented:
@msfroh Do not feel sorry for this.
With 2.9.0 there was another breaking change that impacted me, see https://github.com/opensearch-project/OpenSearch/pull/7508#issuecomment-1658161236

I also do not know what is the solution here. We could introduce something like "internal_API_change" label which will not go to the change log but can be used on GitHub to filter tickets but it would have to be applied consistently and there is no way how to achieve this (except by increasing the work load).
